### ISSUE TITLE
Sparse 24 Linear

### ISSUE THREAD

**Issue by conceptofmind** [facebookresearch/xformers#975]
Hello,

A peer of mine ran the benchmark script on an A100.

Under what conditions should we see the most significant gain for the sparse 24 linear or activations?

```
[-------------------------------------------------------------------------- mlpfwbw --------------------------------------------------------------------------]
                                  |   act24   |   dense   |    w24    |  s24_inp_sparsify24  |  s24_inp_apply_dense  |  s24_inp_apply_dense_t  |  s24_inp_clone
1 threads: ----------------------------------------------------------------------------------------------------------------------------------------------------
      f16 (16384,2048,6144,2048)  |   9871.2  |  10735.6  |   9883.9  |        191.5         |         113.5         |          113.9          |      109.7
      f16 (2048,5120,15360,5120)  |   7558.9  |   7802.9  |  10832.7  |         70.0         |          36.9         |           36.4          |       40.2
      f16 (1024,8192,24576,8192)  |  11489.1  |  10542.3  |  17096.2  |         58.4         |          28.0         |           28.8          |       25.0
      f16 (2048,8192,24576,8192)  |  18698.8  |  19723.3  |  31712.7  |        102.7         |          61.5         |           60.1          |       59.8
      f16 (4096,8192,24576,8192)  |  34372.5  |  38973.1  |  60486.1  |        192.2         |         114.9         |          114.3          |      109.6
      f16 (44160,1024,4096,1024)  |  11072.5  |  10270.5  |   9689.1  |        254.7         |         156.1         |          152.9          |      145.5
      f16 (12288,1536,6144,1536)  |   6040.8  |   6374.0  |   5868.6  |        114.5         |          66.5         |           67.9          |       65.4

Times are in microseconds (us).
```

Thank you,

Enrico

**Comment by danthe3rd** [comment]
Hi @conceptofmind 
There is an overhead to pay to sparsify a GEMM input, but then the sparse GEMM is faster. So the best scenario is when the operand that you sparsify is tiny compared to the other operand (eg if the model is fairly small, but you have a huge batch size, it's better to sparsify the weights).
By default, the sparse GEMM runs with our kernels based on cutlass. We can also use cusparselt - which we found a bit faster - if you follow these steps:
(1) Set `backend="cusparselt"` when you call [`sparsify24`](https://github.com/facebookresearch/xformers/blob/31ef3152a903ca6bbc3dcbf2e51df908fa720c73/xformers/ops/sp24.py#L613)
(2) This only supports sparsifying the weights for now (might support sparsifying activations in some special cases)
(3) You need to have PyTorch with cusparselt (install pytorch nightly from *pip*, as the conda builds didn't have it as of some time ago)

Also if you are training, you can re-use the sparsified matrix in the BW pass - so you pay the cost of sparsification only once. In practice, the FFN part of transformers can be ~20% faster with 2:4 sparsity for the weights (with fairly large batch sizes).

Let me know if you have other questions :)
Also I plan to post something on [twitter](https://twitter.com/d_haziza) next week to explain the principles and how to use it.

**Comment by conceptofmind** [comment]
Hi @danthe3rd ,

Thank you for the insight.

Say we define an FNN with weight sparsity using `backend="cusparselt"`:

```python
class SparseLinear(torch.nn.Linear):
    def forward(self, input: torch.Tensor) -> torch.Tensor:
        w_sparse = xops.sparsify24(
            self.weight,
            gradient="24dense",
            backend="cusparselt",
        )
        return F.linear(input, w_sparse, self.bias)

class FNN(nn.Module):
    def __init__(
        self, 
        dim, 
        hidden_dim, 
    ):
        super().__init__()

        self.linear_in = SparseLinear(dim, hidden_dim,)
        self.act_fn = nn.GELU()
        self.linear_out = SparseLinear(hidden_dim, dim,)

    def forward(self, x):
        x = self.linear_in(x)
        x = self.act_fn(x)
        x = self.linear_out(x)
        return x
```
In this case, the batch size per GPU will be a rough approximate of 64 during DDP training. The FFN will be under 200m parameters.

Is there a specific boundary for the batch size we should be looking for? 32, 64, 128 batch size per GPU? From my understanding, the higher the batch size the more performant sparsity is?

Do you mind providing more information in regards to this statement:
```
Also if you are training, you can re-use the sparsified matrix in the BW pass - so you pay the cost of sparsification only once. In practice, the FFN part of transformers can be ~20% faster with 2:4 sparsity for the weights (with fairly large batch sizes).
```
Do any specific modifications need to be made to the `SparseLinear` to ensure that the sparse matrix is being re-used in the backwards pass?

I greatly appreciate your help.

Thank you,

Enrico

**Comment by danthe3rd** [comment]
> Is there a specific boundary for the batch size we should be looking for

I believe it would be easier for you to just benchmark without 24 sparsity and without. I don't have the numbers from the top of my mind.

> Do any specific modifications need to be made to the SparseLinear to ensure that the sparse matrix is being re-used in the backwards pass during training?

Nope, this is done automatically

> Also, does this work with things like gradient checkpointing or zero redundancy optimizer?

Yes and yes. Although with Zero-type optimizers you will sparsify the weights on all ranks, and this could potentially be avoided as a further optimization. But it will work
